### PR TITLE
Supports externally provided chat log stores

### PR DESCRIPTION
**Pull request by beaucollins**
When moving to a system that runs multiple happychat instances, cached chat logs need to be stored outside of the memory in a single instance, otherwise chat messages originating on _Server A_ will only be in the chat log cache on _Server A_ and connections on other servers will not receive those messages.

## Message Filters

- message filtering middlewares have been renamed te `message-filter` to more clearly describe what they're for
- message filters that block messages from being sent no longer use a rejected promise as the way to indicate a message should be blocked and instead resolve a falsey value. Promise rejections will only occur because of errors thrown in message filters.

## Tests

Tests have been updated to support the new service interface. An improved Socket.IO mock is added to support the correct room API's (`in` and `to`).

### CODE REVIEW

**Comment by beaucollins** [comment]
I can't reproduce the failing test on circle ci. `npm test` on a circle ci shell passes but running under the test coverage report in an ssh shell fail.

File: src/message-filter.js [hunk 1]
```diff
@@ -36,20 +36,18 @@ const runMiddleware = ( provided = [] ) => {
...
 			}
 			// kick off the middleware processing
 			run( { origin, destination, chat, user, message }, context )
 		} )
 		.then( result => {
-			if ( ! result ) {
-				throw new Error( `middleware prevented message(id:${ message.id }) from being sent from ${ origin } to ${ destination } in chat ${ chat.id }` )
-			}
+			// if ( ! result ) {
+			// 	return reject( new Error( `middleware prevented message(id:${ message.id }) from being sent from ${ origin } to ${ destination } in chat ${ chat.id }` ) )
```
Comment (mattwondra) [inline src/message-filter.js:47]: From the PR description, it sounds like this is intentionally commented-out — should we commit and just remove the code instead? That would also let us confidently remove the `reject` parameter added above.

File: src/state/index.js [hunk 2]
```diff
@@ -24,18 +25,19 @@ const selector = evolve( { chatlist: filterClosed } )
 export const serializeAction = () => ( { type: SERIALIZE } )
 export const deserializeAction = () => ( { type: DESERIALIZE } )
 
-export default ( { io, customerAuth, operatorAuth, agentAuth, messageMiddlewares = [], timeout = undefined }, measure = ( key, fn ) => fn ) => {
+export default ( { io, customerAuth, operatorAuth, agentAuth, messageFilters = [], timeout = undefined }, chatLogFactory, measure = ( key, fn ) => fn ) => {
+	const messageFilter = ( ... args ) => run( messageFilters )( ... args );
```
Comment (mattwondra) [inline src/state/index.js:29]: There may be an obvious reason this won't work that I'm missing but — can't we simplify by removing the wrapping function here?

```
const messageFilter = run( messageFilters );
```

File: src/state/middlewares/system/controller.js [hunk 3]
```diff
@@ -11,45 +20,76 @@ import {
...
-		this.maxMessages = options.maxMessages
-		this.chats = {}
+		this.maxMessages = options.maxMessages;
+		this.chats = {};
 	}
 
+	/**
+	 * Appends message to chat of `id`
+	 *
+	 * @param { String } id - chat id related tot message
```
Comment (mattwondra) [inline src/state/middlewares/system/controller.js:57]: #nitpick `related tot message` — assuming should be `related to the message`?

File: src/state/middlewares/system/controller.js [hunk 4]
```diff
@@ -11,45 +20,76 @@ import {
...
+	 * @param { Object } message - message to append to chat log
+	 * @returns { Promise } resolves whet append is completed
```
Comment (mattwondra) [inline src/state/middlewares/system/controller.js:59]: #nitpick `resolves whet append` -> `resolves when the append`

File: src/state/middlewares/system/controller.js [hunk 5]
```diff
@@ -11,45 +20,76 @@ import {
...
 		this.chats = assoc( id, compose(
-			slice( - this.maxMessages, Infinity ),
+			slice( -this.maxMessages, Infinity ),
 			append( message )
-		)( this.findLog( id ) ), this.chats )
+		)( this.getLog( id ) ), this.chats );
+		return Promise.resolve();
 	}
 
+	getLog( id ) {
```
Comment (mattwondra) [inline src/state/middlewares/system/controller.js:69]: The difference between `getLog` and `findLog` isn't very obvious from their names... Any way we could make this clearer?

File: src/state/middlewares/system/controller.js [hunk 6]
```diff
@@ -61,164 +101,211 @@ const formatAgentMessage = ( author_type, author_id, session_id, { id, timestamp
...
 
-	const runMiddleware = ( ... args ) => run( middlewares )( ... args )
+/**
+ * Creates controlled middleware
+ *
+ * @param { Function } messageFilter - list of message filters to be applied to every message
+ * @returns { Function } redux middleware
+ */
+export default ( messageFilter, chatLogFactory = defaultChatLog ) => store => {
+	const cache = { operator: chatLogFactory( 'operator' ), customer: chatLogFactory( 'customer' ) };
```
Comment (mattwondra) [inline src/state/middlewares/system/controller.js:115]: When `chatLogFactory` uses `defaultChatLog`, the parameters `'operator'` and `'customer'` are not used. Is this intentional, assuming that some other consumer of `happychat-service` will want to use those values in their own log factory function?

**Review by mattwondra — COMMENTED** [review]
Left some notes. The code looks sensible, tomorrow morning I'll road test it in-browser for verification that it's all working as expected.

**Pull request closed without merging**